scripts/Diffraction/isis_powder/routines/InstrumentSettings.py

```diff
@@ -0,0 +1,108 @@
+                                 "\nPlease contact the development team.")
+
+    def check_expected_attributes_are_set(self, expected_attr_names):
+        for expected_attr in expected_attr_names:
+            if not [attr_entry for attr_entry in self._attr_mapping if expected_attr == attr_entry[-1]]:
+                raise ValueError("Expected attribute '" + str(expected_attr) + "' is unknown to attribute mapping")
+
+        # Filter down the full mapping list
+        found_tuple_list = [tuple_entry for tuple_entry in self._attr_mapping if tuple_entry[-1] in expected_attr_names]
+        expected_params_dict = dict(found_tuple_list)
+        self._check_attribute_is_set(expected_params_dict)
+
+    def update_attributes(self, advanced_config=None, basic_config=None, kwargs=None, suppress_warnings=False):
+        self._adv_config_dict = advanced_config if advanced_config else self._adv_config_dict
+        self._basic_conf_dict = basic_config if basic_config else self._basic_conf_dict
+        self._kwargs = kwargs if kwargs else self._kwargs
+
+        has_known_keys_already_been_printed = self._unknown_keys_found
+        # Only update if one in hierarchy below it has been updated
+        if advanced_config:
+            self._parse_attributes(self._adv_config_dict, suppress_warnings=suppress_warnings)
+        if advanced_config or basic_config:
+            self._parse_attributes(self._basic_conf_dict,
+                                   suppress_warnings=(not bool(basic_config or suppress_warnings)))
+        if advanced_config or basic_config or kwargs:
+            self._parse_attributes(self._kwargs, suppress_warnings=(not bool(kwargs or suppress_warnings)))
+
+        if not has_known_keys_already_been_printed and self._unknown_keys_found:
+            _print_known_keys(self._attr_mapping)
+
+    def _parse_attributes(self, dict_to_parse, suppress_warnings=False):
+        if not dict_to_parse:
+            return
+
+        for config_key in dict_to_parse:
+            # Recurse down all dictionaries
+            if isinstance(dict_to_parse[config_key], dict):
+                self._parse_attributes(dict_to_parse[config_key])
+                continue  # Skip so we don't accidentally re-add this dictionary
+
+            # Update attributes from said dictionary
+            found_attribute = next((attr_tuple for attr_tuple in self._attr_mapping
+                                    if config_key == attr_tuple[0]), None)
+            if found_attribute:
+                # The first element of the attribute is the config name and the last element is the friendly name
+                self._update_attribute(attr_name=found_attribute[-1], attr_val=dict_to_parse[found_attribute[0]],
+                                       friendly_name=found_attribute[0], suppress_warnings=suppress_warnings)
+            elif not suppress_warnings:
+                warnings.warn("Ignoring unknown configuration key: " + str(config_key))
```

Comment:
Good point I'll ask when I next see them and it can target the PR with initial fixes